tools/restatectl/src/connection.rs

```diff
@@ -0,0 +1,267 @@
+        let mut config_guard = self.nodes_configuration.lock().await;
+        if let Some(config) = &*config_guard {
+            return Ok(config.clone());
+        }
+
+        if self.addresses.is_empty() {
+            return Err(ConnectionInfoError::NoAvailableNodes(NoRoleError(None)));
+        }
+
+        let mut latest_configuration: Option<NodesConfiguration> = None;
+        let mut answer = false;
+        let mut errors = NodesErrors::default();
+        let mut cache = self.cache.lock().await;
+
+        for (address, channel) in self.channels() {
+            let mut client = NodeCtlSvcClient::new(channel.clone())
+                .accept_compressed(CompressionEncoding::Gzip)
+                .send_compressed(CompressionEncoding::Gzip);
+
+            let response = match client.get_ident(()).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address, status);
+                    continue;
+                }
+            };
+
+            // node is reachable and answered
+            answer = true;
+            if response.status != NodeStatus::Alive as i32 {
+                // node did not join the cluster yet.
+                continue;
+            }
+
+            // cache the channel for later usage
+            cache.insert(
+                response
+                    .node_id
+                    .map(|id| PlainNodeId::from(id.id))
+                    .expect("id must be set"),
+                channel,
+            );
+
+            let req = GetMetadataRequest {
+                kind: MetadataKind::NodesConfiguration.into(),
+                sync: false,
+            };
+
+            let mut response = match client.get_metadata(req).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address, status);
+                    continue;
+                }
+            };
+
+            let nodes_configuration =
+                StorageCodec::decode::<NodesConfiguration, _>(&mut response.encoded)
+                    .map_err(|err| ConnectionInfoError::InvalidNodesConfiguration(address, err))?;
+
+            if nodes_configuration.version()
+                > latest_configuration
+                    .as_ref()
+                    .map(|c| c.version())
+                    .unwrap_or(Version::INVALID)
+            {
+                latest_configuration = Some(nodes_configuration);
+            }
+        }
+
+        if !answer {
+            // all nodes has errored
+            return Err(ConnectionInfoError::NodesErrors(errors));
+        }
+
+        *config_guard = latest_configuration.clone();
+        latest_configuration.ok_or(ConnectionInfoError::MissingNodesConfiguration)
+    }
+
+    pub async fn try_each<F, T, Fut>(
+        &self,
+        role: Option<Role>,
+        mut closure: F,
+    ) -> Result<Response<T>, ConnectionInfoError>
```

Comment:
Since we only ever construct either `ClusterCtrlSvcClient` or `NodeCtlSvcClient` maybe it would be useful to just have two convenience wrappers that pass the closure a pre-instantiated client? That would eliminate a fair amount of duplicated code across the various commands.

Reply:
I actually tried that but it's much harder than it sounds because of the compiler complaining about captured variable cannot escape `FnMut` closure body, etc.. so I let it go for now. I will definitely try to do this later.

Reply:
No worries then, let's get this in and we can try again later :-)